provisionersdk/logger_test.go

```diff
@@ -0,0 +1,123 @@
+package provisionersdk_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/provisionersdk"
+	"github.com/coder/coder/provisionersdk/mocks"
+	"github.com/coder/coder/provisionersdk/proto"
+)
+
+type logMatcher struct {
+	level  proto.LogLevel
+	output string
+}
+
+func (m logMatcher) Matches(x interface{}) bool {
+	switch r := x.(type) {
+	case *proto.Provision_Response:
+		return m.logMatches(r.GetLog())
+	case *proto.Parse_Response:
+		return m.logMatches(r.GetLog())
+	default:
+		return false
+	}
+}
+
+func (m logMatcher) logMatches(l *proto.Log) bool {
+	if l.Level != m.level {
+		return false
+	}
+	if l.Output != m.output {
+		return false
+	}
+	return true
+}
+
+func withLog(log *proto.Log) func(x interface{}) bool {
+	m := logMatcher{level: log.GetLevel(), output: log.GetOutput()}
+	return m.Matches
+}
+
+func TestProvisionLogger_Log(t *testing.T) {
+	t.Parallel()
+
+	mStream := new(mocks.ProvisionStream)
+	defer mStream.AssertExpectations(t)
+
+	l := &proto.Log{Level: proto.LogLevel_ERROR, Output: "an error"}
+	mStream.On("Send", mock.MatchedBy(withLog(l))).Return(nil)
+
+	uut := provisionersdk.NewProvisionLogger(mStream)
+	err := uut.Log(l)
+	require.NoError(t, err)
+}
+
+func TestParseLogger_Log(t *testing.T) {
+	t.Parallel()
+
+	mStream := new(mocks.ParseStream)
+	defer mStream.AssertExpectations(t)
+
+	l := &proto.Log{Level: proto.LogLevel_ERROR, Output: "an error"}
+	mStream.On("Send", mock.MatchedBy(withLog(l))).Return(nil)
+
+	uut := provisionersdk.NewParseLogger(mStream)
+	err := uut.Log(l)
+	require.NoError(t, err)
+}
```

Comment:
It seems this log abstraction tests that: a provisioner logger sends a provision log type, and a parse logger sends a parse log type. I feel like this is a lot of code to do that. Would this be caught by many other tests in the codebase?

Reply:
Well, the compiler actually ensures that the types match up.  This tests that a call to Log results in a call to Send and that the fields match up.  Pretty basic, I know, but it's a baseline for the test below for `LogWriter()`.

These small, basic, unit tests are important right by the code they test.  It might be the case that other tests in the codebase actually cover this function, but it breaks encapsulation to depend on that for test coverage; changes elsewhere shouldn't cause us to leave this code exposed.

Reply:
I don't disagree, but the decomposition of this comes at a relatively high cost of maintainability. Adding ~400 lines for logging seems like a *lot* right now (I'm including mock code and tests in that count), considering what we gain from it.

I'm not convinced this abstraction reduces complexity for an implementor and seems premature when we only have a single provisioner emitting logs or using this abstraction. It's hard for me to understand as a contributor the purpose of this abstraction when it's only used in a single place.

Reply:
lines of code is a poor metric for complexity in the first place, and counting autogenerated mocks and test code in your count is counterproductive because it penalizes code that is thoroughly tested versus code that is not.

This abstraction _increases maintainability_ because it separates concerns and keeps logging boilerplate out of the provisioner functions which bloat those functions and makes them harder to read.

Reply:
There is _some basis_ of lines of code to complexity in solving a problem. It's not strong, but if the difference is a significant factor it can be a warning of complexity. This code is ~45 lines of code in `main` (https://github.com/coder/coder/blob/main/provisioner/terraform/provision.go#L150-L195).

There are no tests in the Terraform provider that guarantee logs are sent, so I'd be in favor of moving the testing to there for now and abstracting this once we add providers. I agree with the separation of concerns, but adding an interface that's only been implemented a single time certainly adds cognitive overhead when navigating the codebase.

There are presently no parse logs sent from the Terraform provider, so half of the purpose for this abstraction isn't actually used (apart from tests) at all. It increases maintainability assuming there are many provisioners, but we only have one.

Reply:
I'm happy to get others' perspectives on this too, it's very possible I'm being too pedantic. I care a lot about all of the code having a clear purpose for its structure, hence my hesitation to abstract before multiple usages pop up.

Reply:
I already touched on this a bit in another comment. But I kind of agree with Kyle here, at least some of the abstractions. I'm fine with the `Logger` interface (I'm not sure how we'd do it otherwise), but I think the other logic could be moved into the terraform package. It's usually best to wait for 2+ (more the merrier) use-cases before creating an abstraction, that way it's easier to reason about what the common functionality should be.

And there's no need to be afraid to copy-paste code between packages until enough use-cases surface! As saying GOes:

> [A little copying is better than a little dependency.](https://go-proverbs.github.io)